ostree_upload_server/importers/base.py

```diff
@@ -0,0 +1,84 @@
+import logging
+
+from abc import ABCMeta, abstractmethod
+
+from gi.repository import GLib, Gio
+
+from .util import copy_commit, open_repository, verify_commit_sig
+
+
+class BaseImporter(object):
+    __metaclass__ = ABCMeta
+
+    def __init__(self, src_path, repository_path, gpg_homedir, keyring, sign_key):
+        self._src_path = src_path
+        self._repo_path = repository_path
+        self._gpg_homedir = gpg_homedir
+        self._keyring = keyring
+        self._sign_key = sign_key
+
+    @property
+    def MIME_TYPE(self):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def import_to_repo(self):
+        pass
+
+    @abstractmethod
+    def _import_commit(self, target_repo, src_path_obj, refs=None):
+        pass
+
+    def _apply_commit_to_repo(self, commit, ref):
+        target_repo = open_repository(self._repo_path)
+
+        # Skip the rest of processing if the current ref is the same as in delta
+        _, current_rev = target_repo.resolve_rev(ref, allow_noent=True)
+        logging.debug('Current %s commit: %s', ref, current_rev)
+        if current_rev == commit:
+            logging.info('Ref %s already at commit %s. Skipping changes.', ref, commit)
+            return
+
+        # Apply the delta to our target repo
+        try:
+            src_path_obj = Gio.File.new_for_path(self._src_path)
+
+            # Prepare the transaction
+            target_repo.prepare_transaction(None)
```

Comment:
This should be outside the `try` so you don't abort a transaction that never started.

Reply:
Good point - will fix (though I think that was like that before)